[WZWren] iP

WZWren opened the pull request:
# TrackerBot
Ever forgot something the moment you turned around after you reminded yourself to do it?

We've got you covered! Introducing TrackerBot, the:
 -  [X] Lightweight
 -  [X] Text-Based, and
 -  [X] Easy-to-Use app!
 - [ ] ~~Only downside is, you'd need to keep it up forever...~~
 - [X] **SCRATCH THAT**, it saves your tasks when you quit the app!

Want to try it out?
 1. Go to the [Release page](https://github.com/WZWren/ip/releases/tag/A-Jar)...
 1. Download the .jar file...
 1. Place it in an empty folder of your choice...
 1. ...and run it! Easy as pie. 🥧

> “Memory is the treasury and guardian of all things.” ~ Cicero

Just leave the hard parts to us, and use the rest of that space on something you'd treasure.

## What's inside?

- Ability to add 3 types of task classification, with room to expand.
- Mark, unmark and delete these tasks.
- Save and load into the jar file folder automatically
- Portable file - bring it anywhere, use it anytime.

Not interested in a jar file? Download the source code and create an entry point:
```
public class Main {
    public static void main(String[] args) {
        new TrackerBot("your custom app name here").run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

WZWren committed:
Renamed the Duke file to TrackerBot.

WZWren committed:
Added Greet to TrackerBot.

WZWren committed:
Added Exit to TrackerBot.

WZWren committed:
Added input handling to TrackerBot.

WZWren committed:
Added variables to store tasks.

WZWren committed:
Added List, Add methods to TrackerBot.

WZWren committed:
Updated version in Javadoc.

WZWren committed:
Added Task class to project.

The Task class encapsulates the name and mark of a reminder task - this allows us to more easily store our tasks in the future, and has the immediate benefit of storing the completion status of each individual item in the list.

WZWren committed:
Converted Task List to use the Task class.

WZWren committed:
Added code for marking/unmarking Tasks.

WZWren committed:
Modified input handler to prepare for Scanner use.

WZWren committed:
Linked mark, unmark to main.

WZWren committed:
Added Todo, Deadline and Event classes.

WZWren committed:
Added utility DS Pair to the project.

WZWren committed:
Added an enum for Commands.

This is done to reduce future boiler code and simplify logic in the TrackerBot class.

WZWren committed:
Implemented Enum into handler logic.

WZWren committed:
Adjusted Event toString output for formatting.

WZWren committed:
Changed add logic for Task for variants.

WZWren committed:
Changed Task to abstract class.

WZWren committed:
Fixed toString of Event to properly tag the String

WZWren committed:
Converted Task array to ArrayList, fixed getTask.

getTask had a minor OutOfBounds error that could occur - as this was caught during conversion and the ArrayList bug fix is minor, this fix was kept together with the conversion commit.

WZWren committed:
Added Delete to TrackerBot

WZWren committed:
Modified input, runtest.bat files to fit new name of project.

WZWren committed:
Modified expected file to match the expected output.

User input is entered after a print(...) line instead of a println(...) line, and does not have a newline inside of it. As a result, the test will have divider lines overlapping with the user console. This is not very desirable behavior in the use case where users redirect their input into the program, but as this app is primarily a console-based program, we will not fix this yet.

WZWren committed:
Fixed bug where error message displayed for event is incorrect.

WZWren committed:
Fix indent and switch cases

WZWren committed:
Add skeleton code for save string in Task

The method to parse the save string is currently unimplemented.
We will only implement it after creating the Factory method for Tasks.

WZWren committed:
Add skeleton code for factory method of Task

WZWren committed:
Add basic save method with IO output

WZWren committed:
Add basic IO read from file

WZWren committed:
Add abstract saveString generator for Task

We separate getSaveInfo from the child classes to maintain the
abstraction barrier between Task and its child classes, while still
making it compulsory to override toSaveString in the child class.

WZWren committed:
Add factory method logic in Task

WZWren committed:
Add save/load logic for Task

WZWren committed:
Merge branch 'branch-Level-7'

WZWren committed:
Add String to Date conversion method

WZWren committed:
Move conversion method to new class

WZWren committed:
Change String fields to LocalDateTime

WZWren committed:
Change logic for save file date handling

WZWren committed:
Merge branch 'branch-Level-8'

WZWren committed:
Move CommandType enum out of main

WZWren committed:
Group task methods from main into Storage

WZWren committed:
Add skeleton code for UI

WZWren committed:
Move user input ID to Parser

WZWren committed:
Move IO methods to Storage

WZWren committed:
Add TrackerBotException class

TrackerBot used to manually set exceptional cases and print them
out onto the console.

The change to a more OOP approach made it more difficult to perform
this approach while abiding by encapsulation approach.

Adding the Exception class allows us to push the error handling to
a more appropriate class, simplifying code.

WZWren committed:
Move Task parse logic to Parser

WZWren committed:
Change main error handling to use Exception

WZWren committed:
Rename Storage to TaskList

TaskList better represents what the class should be doing.

WZWren committed:
Add Command class

TrackerBot currently handles each individual Command case using
a switch case and an enum.

As we switch over to a more rigid OOP structure, it becomes unclear
where this handling should be made.

By extracting the Command class, we can handle it all in the
Command class, and call command.execute().

WZWren committed:
Move UI elements to Ui

WZWren committed:
Adjust Parser, TaskList return types

WZWren committed:
Port command logic to Command

WZWren committed:
Add Storage class, minor fixes to UI

WZWren committed:
Change main to be entry, not program

WZWren committed:
Merge branch 'branch-A-MoreOOP'

WZWren committed:
Remove unused Pair class

WZWren committed:
Rename package ip to trackerbot

WZWren committed:
Move all classes into packages

WZWren committed:
Merge branch 'branch-A-Packages'

WZWren committed:
Merge branch 'add-gradle-support'

WZWren committed:
Add gradle support for TrackerBot

WZWren committed:
Merge branch 'master' into branch-A-Packages

WZWren committed:
Refactor package layout

WZWren committed:
Merge branch 'branch-A-Packages'

WZWren committed:
Add test files for Parser, DateHandler

WZWren committed:
Update build.gradle for JAR file

WZWren committed:
Add skeleton code for Find Command

WZWren committed:
Add Javadocs for Command

WZWren committed:
Fix coding standard in Command

The Javadoc comments were added in to fix the checkstyle issue.
While they do not fulfil the minimum requirements for proper
documentation, as we are doing 3 branches in parallel, one of which
is for Javadoc, we will resolve this during the merge conflict step.

WZWren committed:
Add logic for search in Task, TaskList

WZWren committed:
Add Javadocs for CommandType, exceptions

WZWren committed:
Fix styling for Command

WZWren committed:
Fix styling for exception package

WZWren committed:
Fix styling for task package

WZWren committed:
Add Find logic to Command

WZWren committed:
Merge branch 'branch-Level-9'

WZWren committed:
Merge branch 'master' into branch-A-JavaDoc

WZWren committed:
Merge branch 'master' into branch-A-CodingStandard

# Conflicts:
#	src/main/java/trackerbot/command/CommandType.java
#	src/main/java/trackerbot/task/TaskList.java

WZWren committed:
Update Javadoc for command, exception

WZWren committed:
Add Javadoc for task package

WZWren committed:
Fix naming convention on find method

WZWren committed:
Update coding style for rest of code

WZWren committed:
Add Javadoc for rest of the app

WZWren committed:
Update import ordering in test

WZWren committed:
Merge branch 'branch-A-JavaDoc'

WZWren committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/trackerbot/command/Command.java
#	src/main/java/trackerbot/command/CommandType.java
#	src/main/java/trackerbot/exception/TrackerBotException.java
#	src/main/java/trackerbot/task/Task.java
#	src/main/java/trackerbot/task/TaskList.java
#	src/main/java/trackerbot/utils/Storage.java
#	src/main/java/trackerbot/utils/TaskDateHandler.java
#	src/main/java/trackerbot/utils/Ui.java

WZWren committed:
Merge branch 'master' into branch-A-JavaDoc

WZWren committed:
Move @see blocks in Javadoc

WZWren committed:
Merge branch 'branch-A-JavaDoc'

itsNatTan commented on src/main/java/trackerbot/task/Task.java:
```diff
@@ -0,0 +1,155 @@
+     * @throws IllegalArgumentException if the save file is corrupted.
+     * @throws DateTimeParseException if the DateTime field is invalid.
+     */
+    public static Task ofSaveString(String type, String... args)
```
Good checks for invalid arguments

itsNatTan commented on src/test/java/trackerbot/utils/ParserTest.java:
Very comprehensive tests

itsNatTan reviewed (commented):
Very good JavaDocs and version tracking. No complaints 11/10

chuababyy commented on src/main/java/trackerbot/command/Command.java:
```diff
@@ -0,0 +1,226 @@
+            ui.showMessage(tasks.add(type, commandFields));
+        }
+
+        public boolean isExit() {
```
I noticed that the `@Override` notation for `execute` and `isExit` function is not included for all the child classes. It would be good to include this notation so that any errors can be caught by the compiler at compile time!

WZWren replied:
Ah, I must have forgotten about this. Thanks for pointing it out!

chuababyy commented on src/main/java/trackerbot/command/Command.java:
```diff
@@ -0,0 +1,226 @@
+        }
+
+        Command result;
+        switch (parsedType) {
```
Good that you included the explicit `// Fallthrough` comment in accordance with the Java coding standard in our course

chuababyy commented on src/main/java/trackerbot/task/Task.java:
```diff
@@ -0,0 +1,155 @@
+     * @return true if the task is successfully marked to completion,
+     *         false if the task is already completed.
+     */
+    public boolean markTask() {
+        if (isDone) {
+            return false;
+        }
+        isDone = true;
+        return true;
+    }
+
+    /**
+     * Flags the task to be incomplete, if able.
+     * @return true if the task is successfully marked to be incomplete,
+     *         false if the task is already marked as incomplete.
+     */
+    public boolean unmarkTask() {
+        if (!isDone) {
+            return false;
+        }
+        isDone = false;
+        return true;
+    }
```
Given that this is returning a boolean, it would be good to name it to sound like a boolean in accordance to the Java coding standard by our course. 

```suggestion
    public boolean canMarkTask() {
        if (isDone) {
            return false;
        }
        isDone = true;
        return true;
    }

    /**
     * Flags the task to be incomplete, if able.
     * @return true if the task is successfully marked to be incomplete,
     *         false if the task is already marked as incomplete.
     */
    public boolean canUnmarkTask() {
        if (!isDone) {
            return false;
        }
        isDone = false;
        return true;
    }
```

WZWren replied:
Hmm, should this really be named this way? These methods actually mark/unmark the task directly, and the boolean flag is actually to feedback to the caller whether the task has been marked/unmarked successfully.

chuababyy commented on src/main/java/trackerbot/utils/Ui.java:
```diff
@@ -0,0 +1,87 @@
+ */
+public class Ui {
+    /** Line separators for the console between paragraphs. **/
+    private static final String FORMAT_LINE =
```
Good that you named your variables in accordance to the Java coding standards given in our course! Since these variables are constants, they are named in all uppercase using underscore to separate words

WZWren committed:
Add dependencies to JavaFX

WZWren committed:
Add skeleton code from JFX tutorial

WZWren committed:
Add images for JFX

WZWren committed:
Convert Ui to UiHandler

Ui is the direct handler class for the text-based UI.

With the addition of the GUI, there is no need for the Ui to handle
actions anymore.

UiHandler is an experimental class to see if the Ui abstraction can
hold for a GUI setup.

WZWren committed:
Add entry point to JFX

WZWren committed:
Fix to entry point in FXML

WZWren committed:
Add WIP code for UiHandler

WZWren committed:
Rename methods to match project

WZWren committed:
Update gradle entry point

WZWren committed:
Remove deprecated methods

WZWren committed:
Add delay before exit

WZWren committed:
Add aesthetic changes to app

WZWren committed:
Update checkstyle compliance

WZWren committed:
Merge branch 'branch-Level-10'

rocketninja7 commented on src/main/java/trackerbot/command/Command.java:
```diff
@@ -0,0 +1,226 @@
+
+        Command result;
+        switch (parsedType) {
+        case TODO:
```
I like that your case statements have the correct indentation.

rocketninja7 commented on src/main/java/trackerbot/gui/DialogBox.java:
```diff
@@ -0,0 +1,94 @@
+        dialog.setText(text);
+        displayPicture.setImage(img);
+
+        Circle circleClip = new Circle(
```
I like that your statements here have indentation of 8 spaces and that the lines are split appropriately.

rocketninja7 commented on src/main/java/trackerbot/task/Task.java:
```diff
@@ -0,0 +1,155 @@
+     * @return true, if the description contains the searchStr, and false
+     *         otherwise.
+     */
+    public boolean doesContain(String searchStr) {
```
I like how you name methods that return booleans according to the standards.

rocketninja7 commented on src/test/java/trackerbot/utils/TaskDateHandlerTest.java:
```diff
@@ -0,0 +1,75 @@
+ */
+public class TaskDateHandlerTest {
+    @Test
+    public void convertInputToDate_dayAndMonthInput_shouldParseProperly() throws DateTimeParseException {
```
I like how you name your test methods.

rocketninja7 reviewed (commented):
Excellent work!

WZWren committed:
Fix minor DialogBox bug

WZWren committed:
Extract nested subclasses of Command

Currently, Command nests all the subtypes of command as part of itself.
This allows the abstract Command Class to conveniently return its
sub-types using the factory method.

However, this was done when the classes have not been properly packaged,
where having all the classes be in 1 folder would have bloated the
project environment significantly. The tradeoff for this would be
causing the Command class itself to bloat unnecessarily.

Therefore, we choose to extract the subclasses inside Command into the
trackerbot.command package.

To maintain the original design philosophy of having only Command be
able to access these classes, we set the relevant access modifiers to
package-private.

WZWren committed:
Enable assertions in Gradle

WZWren committed:
Add assertion for Event class

This assertion exposes a bug for event where from and to does not
maintain a proper relationship.

WZWren committed:
Change exception type from Java-native types

Exceptions are currently mainly handled by NumberFormatException,
DateTimeParseException and TrackerBotException types, with the unchecked
exception IllegalStateException handling irrecoverable states in the
program.

The current method of handling makes it inconvenient to implement new
methods related to methods that throw the Java exception types, as
it is unclear whether the method should throw the native exception or
the specific Java exceptions. Furthermore, it complicates exception
handling downstream, as these specific scenarios need to be handled in
a way that users will be able to understand and rectify.

Therefore, we preemptively catch the Java exceptions and convert them
into a relevant TrackerBotException, that our methods downstream will be
able to handle generically.

This will allow us to program methods more generically in the long run,
and also let linters catch unhandled exceptions that we expect the user
to be able to create during the runtime of the program. In context of
the current branch, it lets us fix the assertions bug with an abstracted
method while also keeping specificity of the error to be thrown.

WZWren committed:
Update JavaDoc, coding standard

WZWren committed:
TaskList: Extract method, Rename method

WZWren committed:
Standardize SLAP in TrackerBot, Command

WZWren committed:
Parser: Extract method from parseAdd

WZWren committed:
Fix bug exposed by Event assertion

WZWren committed:
Fix bug where saves do not purge on error

WZWren committed:
Fix JavaDoc header

WZWren committed:
Extract magic values for URI, format

WZWren committed:
Merge pull request #2 from WZWren/branch-A-CodeQuality

Improve code quality

WZWren committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/trackerbot/utils/Parser.java
#	src/main/java/trackerbot/utils/TaskDateHandler.java

WZWren committed:
Merge pull request #1 from WZWren/branch-A-Assertions

Add assertions for Events

WZWren committed:
Add skeleton code for mass ops

WZWren committed:
Create MassCommand subtype

WZWren committed:
Add parser for mass operation strings

WZWren committed:
Fix parser logic, add command logic

WZWren committed:
Add logic for mass deletion

WZWren committed:
Add logic for mass toggle

WZWren committed:
Change mass delimiter to single character

The Mass command uses double colons "::" to delimit each subsequent
command of the same subtype.

For a command designed to shorthand mass operations, this is very
cumbersome to have to SHIFT+";" twice. Additionally, there is no
real reason to use double colons in the code, as colons and
semicolons do not have a special meaning in the parser.

Therefore, we adjust the delimiter to use a single semicolon, which
should make the process less cumbersome.

WZWren committed:
Add logic for mass search

WZWren committed:
Add exceptional case for mass list

WZWren committed:
Fix accessor for MassCommand class

WZWren committed:
Merge branch 'branch-C-MassOps'

WZWren committed:
Expand MainWindow layout

WZWren committed:
Upload Ui.png

WZWren committed:
Update README.md

WZWren committed:
Adjust comment formatting

WZWren committed:
Remove unnecessary this keyword

WZWren committed:
Merge branch 'branch-CommentFixes'
